### PR TITLE
arch-x86: Add XCR0 register and add to X86KvmCPU

### PR DESCRIPTION
**Pull request by abmerop**
The extended control registers were not being updated in the KVM thread context nor updated in the KVM state. This was causing issues when checkpointing since the XCR0 value was reverting to the default value rather than what it was previously before the checkpoint. THis was causing multiple applications to crash due to executing instructions which are now illegal instructions due to XCR0 being incorrect.

This commit adds the XCR0 as a misc register similar to the exiting x86 control registers and adds all of the helper functions to access and set the register value. It also adds support for updating the KVM CPU's state with the register value and updating the thread context's misc reg value so that it is checkpointed along with the other misc regs.

Note that this does *not* add support for XSAVE of the AVX state (i.e., the upper 128 bits of YMM registers). It does however fix the immediate problem in issue #958 .

Change-Id: I97456c8b57cbc7b381bd4be94944ce6567a43c76

### CODE REVIEW

**Comment by abmerop** [comment]
@v-ramadas I stumbled upon a solution to this kernel panic while working on something slightly related, so created a PR for it.  We should be able to test checkpointing GPU ML apps now

**Comment by v-ramadas** [comment]
Thanks! This is great. I will test it out when I get some time.

File: src/arch/x86/fs_workload.cc
Comment (mattsinc) [inline src/arch/x86/fs_workload.cc]: Ideally it would be good to explain why it must be 1

File: src/arch/x86/regs/misc.hh
Comment (mattsinc) [inline src/arch/x86/regs/misc.hh:408]: it's a little weird that you declare XcrBase, then set it to the Xcr0 in the next line, but I guess if it works ok

**Review by mattsinc — CHANGES_REQUESTED** [review]
Amazing!  Just one small thing,

File: src/arch/x86/regs/misc.hh
Comment (abmerop) [reply src/arch/x86/regs/misc.hh:408]: Just following the style above with the other groups of registers.  There just happens to only be one XCR register

File: src/arch/x86/fs_workload.cc
Comment (abmerop) [reply src/arch/x86/fs_workload.cc]: Added comments

**Comment by abmerop** [comment]
@BobbyRBruce @Harshil2107  The failing tests are checkpoint restore checks.  The checkpoint appears to be part of a resource, how does that get updated?  This change implicitly adds another element to an array param

**Review by mattsinc — DISMISSED** [review]

**Comment by abmerop** [comment]
.... Do I need to write a checkpoint upgrader?

**Comment by Harshil2107** [comment]
If the checkpoint upgrader needs to be updated, then I can run the updated checkpoint upgrader on the existing checkpoints and push a new version. If the checkpoint upgrader doesnt need to be update I can just retake the checkpoints and update the version.

Can you tell me if you need to update the checkpoint upgrader?

**Comment by abmerop** [comment]
Well if you take new checkpoint *with this change applied* then that would also fix the failing tests.

I suppose writing an upgrader is more correct as other might need to use it.  Is the upgrader run as part of the tests?

**Comment by abmerop** [comment]
> If the checkpoint upgrader needs to be updated, then I can run the updated checkpoint upgrader on the existing checkpoints and push a new version. If the checkpoint upgrader doesnt need to be update I can just retake the checkpoints and update the version.
> 
> Can you tell me if you need to update the checkpoint upgrader?

Added a checkpoint upgrader.  I guess this would have to be applied to the resources?

**Comment by Harshil2107** [comment]
I will run and update the failing checkpoints.

**Comment by abmerop** [comment]
> I will run and update the failing checkpoints.

Thank you!

**Comment by Harshil2107** [comment]
I just reran the tests with updated checkpoints.

**Comment by Harshil2107** [comment]
To get the tests to pass you would need to update the resource version to 3.0.0 in the x86-fs-restore-checkpoint.py, x86-hello-restore-checkpoint.py. I can also push a commit with this changes if you want me to.

**Comment by abmerop** [comment]
> To get the tests to pass you would need to update the resource version to 3.0.0 in the x86-fs-restore-checkpoint.py, x86-hello-restore-checkpoint.py. I can also push a commit with this changes if you want me to.

Sure... I don't know where those files are.

**Comment by Harshil2107** [comment]
The checkpoint upgrader doesnt seem to fix the checkpoints, should I take new checkpoints or do we need to update the updater again?

**Comment by abmerop** [comment]
It looks like the testlib tests have a different name for the ini config section, so I will have to rewrite the upgrader

**Comment by abmerop** [comment]
Updated the checkpoint upgrader.  You might need to manually edit the m5.cpt file to remove the x86-add-xcr0 tag from the file before running the upgrader again.  Or simply use the 2.0.0 files

**Comment by abmerop** [comment]
If I could make a suggestion.  Is it possible to have testlib run the checkpoint upgrader before running a test? Then you could "officially" recreate during the release freeze? That way it wouldn't require someone with access to the checkpoint files to post a PR requiring checkpoint upgrades.  Although that might be rare enough it doesn't make sense.

**Comment by Harshil2107** [comment]
The updated simpoint is causing other PRs to fail, I will create create a version specific resource for this PR and so that the tests in other PRs dont fail.

**Comment by Harshil2107** [comment]
> If I could make a suggestion. Is it possible to have testlib run the checkpoint upgrader before running a test? Then you could "officially" recreate during the release freeze? That way it wouldn't require someone with access to the checkpoint files to post a PR requiring checkpoint upgrades. Although that might be rare enough it doesn't make sense.

The reason for not doing that is, that the checkpoints on gem5 resources become outdated and need to be updated all at once. As we have to keep the checkpoints in sync with gem5 resources, I feel it is better to keep on updating them as the changes are made. 

I will push a change that defines the resource versions explicitly instead of grabbing the latest version in another PR. In this way the tests for other PRs with older code will not fail, when we update the checkpoints.

**Comment by Harshil2107** [comment]
I have created a new resource for this change, now the tests in other PRs should not be affected. I will sync the resources when I push the PR I mentioned above (this should not affect tests for other PRs).

**Comment by abmerop** [comment]
> > If I could make a suggestion. Is it possible to have testlib run the checkpoint upgrader before running a test? Then you could "officially" recreate during the release freeze? That way it wouldn't require someone with access to the checkpoint files to post a PR requiring checkpoint upgrades. Although that might be rare enough it doesn't make sense.
> 
> The reason for not doing that is, that the checkpoints on gem5 resources become outdated and need to be updated all at once. As we have to keep the checkpoints in sync with gem5 resources, I feel it is better to keep on updating them as the changes are made.
> 
> I will push a change that defines the resource versions explicitly instead of grabbing the latest version in another PR. In this way the tests for other PRs with older code will not fail, when we update the checkpoints.

I see. I was attempting to find a solution that didn't require a single entity (UC Davis) to have to be involved.

**Comment by powerjg** [comment]
> I see. I was attempting to find a solution that didn't require a single entity (UC Davis) to have to be involved.

Thanks! I agree that would be great. The main problem here is that checkpoints are deeply tied to the code of gem5 (which is a problem). Really, checkpoints should be more of an API that can't be changed willy-nilly, but that's not how it's architected right now.

There's a tradeoff in the work to do per-commit and per-release. When possible, I'd like to do the work per-commit to make release time easier. Updating checkpoints should be rare enough (a few times every release) that it's better to do it per-commit than waiting until the release.

You're not the first to suggest auto-running the checkpoint upgrader (or removing checkpointing from testing). But, one of the main arguments against this is that it would mask when checkpoints need to be updated for the next release (putting more onus on us at release time).

**Comment by mattsinc** [comment]
@abmerop it doesn't look like you need me for anything here at this point, since you are still working with @Harshil2107 on the checkpoint changes.  Let me know if I'm missing something though.  I added Harshil as a review to hopefully help with this.

**Comment by abmerop** [comment]
> @abmerop it doesn't look like you need me for anything here at this point, since you are still working with @Harshil2107 on the checkpoint changes. Let me know if I'm missing something though. I added Harshil as a review to hopefully help with this.

The checks has passed now, so I just need a reviewer to approve

**Review by mattsinc — APPROVED** [review]
Probably @Harshil2107 or @powerjg should approve too?

**Review by Harshil2107 — APPROVED** [review]
lgtm

**Pull request merged**
